applications/StructuralMechanicsApplication/custom_response_functions/README.md

```diff
@@ -0,0 +1,241 @@
+- ```element_sensitivity_variables```: Here are sensitivities with respect to the properties of the elements computed. For that the respective name of the Kratos-Variable has to be given (e.g. ```THICKNESS```, ```I22``` or ```YOUNG_MODULUS```)
+- ```condition_sensitivity_variables```: Here are sensitivities with respect to the properties of the conditions computed. Currenty there is only ```POINT_LOAD``` availible.
+
+**Important, please note:**
+In order to use an element or condition design variable one has to ensure that a corresponding Kratos-Variable for the sensitivity is defined (e.g. for the design variable ```THICKNESS``` a corresponding variable called ```THICKNESS_SENSITIVITY``` is necessary). This additional variable is necessary to store the results of the sensitivity analysis.
+
+There are currently three different types of response functions availible which can be chosen as ```response_type```. For each of them specific settings are neccessary:
+- ```adjoint_nodal_displacement```: The response is the displacement or rotation of a single node. Necessary additional settings are:
+    * ```traced_node```: ID of the traced node
+    * ```traced_dof```: Define the traced DOF (e.g. ```DISPLACEMENT_Z``` or ```ROTATION_X```)
+
+- ```adjoint_strain_energy```: The response is the linear strain energy. No additional settings are necessary.
+
+- ```adjoint_local_stress```: The response is the stress or stress-resultant of a single element. Necessary additional settings are:
+    * ```traced_element```: ID of the element which should be traced
+    * ```stress_type```: Stress type which should be traced (e.g. FX, MY, FXX or MXX)
+    * ```stress_treatment```: There are three possibilities: ```node``` (Takes the response value at the position of a defined node of the element. Only availible for beam element.), ```GP``` (Takes the response value at a defined Gauss-Point of the element.) and ```mean``` (The response is the mean value of all Gauss-Point results of the traced stress type.)
+    * ```stress_location```: Only necessary if ```node``` or ```GP``` is chosen as ```stress_treatment```. Define here the local ID of the position where the stress has to be traced (e.g. if the stress resultant of a beam element should be traced at one of his two nodes ```stress_location``` has to be 1 or 2)
+
+Examples:
+```python
+    "response_function_settings" : {
+            "response_type"     : "adjoint_nodal_displacement",
+            "gradient_mode"     : "semi_analytic",
+            "sensitivity_model_part_name" : "Parts_Beam",
+            "nodal_sensitivity_variables"  : ["SHAPE_SENSITIVITY"],
+            "element_sensitivity_variables"  : ["I22"],
+            "condition_sensitivity_variables"  : ["POINT_LOAD"],
+            "step_size"         : 1e-6,
+            "traced_node"       : 6,
```

Comment:
That's true, however we start to get many levels on the response functions, especially in an optimization context.
This is why i would prefer to not introduce a new level here at the moment.

also the ```gradient_mode``` and ```step_size``` of the perturbation (i cannot find your specific comment anymore) is dependent on the elements, not on the response function - also there you are right.
However currently we also have some response functions on the master, e.g. for eigenfrequencies that do need this setting in the response function settings, because they do not use any adjoint elements, and the finite differencing is done directly in the response function.
So i would like to avoid to have this setting on completely different positions for two different responses.

Reply:
ok